src/source/Ice/SocketConnection.c

```diff
@@ -150,16 +154,23 @@ STATUS socketConnectionSendData(PSocketConnection pSocketConnection, PBYTE pBuf,
     CHK(pSocketConnection != NULL && pBuf != NULL, STATUS_NULL_ARG);
     CHK(bufLen != 0 && (pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP || pDestIp != NULL), STATUS_INVALID_ARG);
     CHK(!pSocketConnection->connectionClosed, STATUS_SOCKET_CONNECTION_CLOSED_ALREADY);
-    CHK_STATUS(socketConnectionReadyToSend(pSocketConnection, &socketReady));
-    CHK_WARN(socketReady, STATUS_SOCKET_CONNECTION_NOT_READY_TO_SEND, "Socket connection not ready to send data");
+    CHK_STATUS(socketConnectionReadyToSend(pSocketConnection, &connected));
+    CHK_WARN(connected, STATUS_SOCKET_CONNECTION_NOT_READY_TO_SEND, "Socket connection not ready to send data");
 
     MUTEX_LOCK(pSocketConnection->lock);
     locked = TRUE;
 
     if (pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP && pSocketConnection->secureConnection) {
         // underlying BIO has been bound to socket so SSL_write sends the encrypted data.
-        sslRet = SSL_write(pSocketConnection->pSsl, pBuf, bufLen);
-        CHK_WARN(sslRet > 0, retStatus, "%s", ERR_error_string(SSL_get_error(pSocketConnection->pSsl, sslRet), NULL));
+        // because we are using non-blocking socket, SSL_write may return SSL_ERROR_WANT_WRITE in which case
+        // SSL_write should be called again with the same buffer.
+        while(iterate) {
+            sslRet = SSL_write(pSocketConnection->pSsl, pBuf, bufLen);
+            if (sslRet > 0 || (sslErr = SSL_get_error(pSocketConnection->pSsl, sslRet)) != SSL_ERROR_WANT_WRITE) {
+                iterate = FALSE;
+            }
+        }
+        CHK_WARN(sslRet > 0, retStatus, "SSL_write failed with %s", ERR_error_string(sslErr, NULL));
```

Comment:
So, this is the case when sslRet == 0 right?

Reply:
not sure what you mean here. if sslRet == 0 it should also fail since socket is already closed.

Reply:
This is an interesting case. The manual states that the return is success if it's > 0 and the value is the number of bytes written. However, in case of a non-block socket it could return 0 and set the SSL_ERROR_WANT_WRITE in error. Perhaps, the loop should be loop until > 0 && err != SSL_ERROR_WANT_WRITE.

Then, after the loop you could check the err itself perhaps?

Reply:
if > 0 it's a success and there would be no error. if the return is 0, it would check if the err is SSL_ERROR_WANT_WRITE, if so retry, otherwise stop iterating. Then CHK_WARN would print the warning.